src/Adapter/Product/FilterParametersUpdater.php

```diff
@@ -55,72 +55,49 @@ public function setPositionOrdering($filterParameters, $orderBy, $hasCategoryFil
-     *
-     * @param array $filterParameters
-     * @param string $offset
-     * @param string $limit
-     * @param string $orderBy
-     * @param string $sortOrder
-     *
+     * @param array $queryFilterParameters
+     * @param array $persistedFilterParameters
+     * @param array $defaultFilterParameters
      * @return array
      */
-    public function setValues(
-        array $filterParameters,
-        $offset,
-        $limit,
-        $orderBy,
-        $sortOrder
+    public function buildFilters(
+        array $queryFilterParameters,
+        array $persistedFilterParameters,
+        array $defaultFilterParameters
     ) {
         return [
-            'offset' => $this->getOffset($offset, $filterParameters),
-            'limit' => $this->getLimit($limit, $filterParameters),
-            'orderBy' => $this->getOrderBy($orderBy, $filterParameters),
-            'sortOrder' => $this->getSortOrder($sortOrder, $filterParameters),
+            'offset' => (int) $this->getParameter('offset', $queryFilterParameters, $persistedFilterParameters, $defaultFilterParameters),
+            'limit' => (int) $this->getParameter('limit', $queryFilterParameters, $persistedFilterParameters, $defaultFilterParameters),
+            'orderBy' => (string) $this->getParameter('orderBy', $queryFilterParameters, $persistedFilterParameters, $defaultFilterParameters),
+            'sortOrder' => (string) $this->getParameter('sortOrder', $queryFilterParameters, $persistedFilterParameters, $defaultFilterParameters),
         ];
     }
 
     /**
-     * @param int $offset
-     * @param array $filterParameters
-     *
-     * @return int
-     */
-    private function getOffset($offset, array $filterParameters)
-    {
-        return ($offset === 'last' && isset($filterParameters['last_offset'])) ? $filterParameters['last_offset'] : $offset;
-    }
-
-    /**
-     * @param int $limit
-     * @param array $filterParameters
-     *
-     * @return int
+     * @param string $parameterName
+     * @param array $queryFilterParameters
+     * @param array $persistedFilterParameters
+     * @param array $defaultFilterParameters
+     * @return string|int
      */
-    private function getLimit($limit, array $filterParameters)
-    {
-        return ($limit === 'last' && isset($filterParameters['last_limit'])) ? $filterParameters['last_limit'] : $limit;
-    }
+    private function getParameter(
+        $parameterName,
+        array $queryFilterParameters,
+        array $persistedFilterParameters,
+        array $defaultFilterParameters
+    ) {
+        if (!empty($queryFilterParameters) && isset($queryFilterParameters[$parameterName])) {
+            $value = $queryFilterParameters[$parameterName];
+        } else if (!empty($persistedFilterParameters) && isset($persistedFilterParameters[$parameterName])) {
+            $value = $persistedFilterParameters[$parameterName];
+        } else {
+            $value = $defaultFilterParameters[$parameterName];
```

Comment:
You must be sure `$defaultFilterParameters[$parameterName]` exists.

Reply:
Yeah I hesitated to let PHP throw an error.. but I'm gonna throw a nice Exception instead ^^